New Variable: Total grounding line discharge entering adjacent ocean cell

alexolinhager opened the pull request:
This PR adds a small subroutine to the hydro model that calculates a new variable, `totalGroundingLineDischargeCell`, that is needed to couple the hydro model to MPAS-Ocean. `totalGroundingLineDischargeCell` is the total volume flux (distributed + channelized; units: m3 s-1) entering each ocean cell adjacent to the grounding line. Distributed and channelized flow at each grounding line edge is summed into `totalGroundingLineDischargeEdge` and that value is extrapolated to the adjacent ocean cell. If multiple grounding line edges border the same ocean cell, then `totalGroundingLineDischargeEdge` values at all adjacent edges are summed to form `totalGroundingLineDischargeCell` for that cell.

alexolinhager commented:
The PR passes the full integration suite at commit `57a370e`

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -1718,7 +1733,6 @@ subroutine update_channel(block, err)
       ! local variables
       !-----------------------------------------------------------------
       ! Pools pointers
-!!      type (mpas_pool_type), pointer :: geometryPool
```
There is a long series of changes to this subroutine that are just variable declarations and mpas_pool_get calls, but no actual changes to the operations - those should all be removed from the PR.

matthewhoffman replied:
@alexolinhager , looks like this series of changes is still in the PR.  Let me know if you need help figuring out how to get rid of them.

alexolinhager replied:
@matthewhoffman I thought I got rid of all extra changes in that subroutine. All of the extra variable names have been deleted from what I can tell. I had accidentally removed `nEdgesSolve` then had to add it back in, which is why that is there as a change now. The remaining changes were me cleaning up the script and removing blank lines left after removing the extra variable names. Are there more changes I should get rid of?

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2271,4 +2303,96 @@ subroutine calc_hydro_mask(domain)
+         cell2 = cellsOnEdge(2, iEdge)
+         
+         if (hydroMarineMarginMask(iEdge) == 1) then
+            ! We are looking for edges with 1 cell grounded ice and the
+            ! other cell floating ice or open ocean
```
this comment doesn't make sense here

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2271,4 +2303,96 @@ subroutine calc_hydro_mask(domain)
+         if (hydroMarineMarginMask(iEdge) == 1) then
+            ! We are looking for edges with 1 cell grounded ice and the
+            ! other cell floating ice or open ocean
+            if ( (li_mask_is_grounded_ice(cellMask(cell1))) .and. &
+                 (li_mask_is_floating_ice(cellMask(cell2)) .or. &
+                 ((bedTopography(cell2) < config_sea_level) .and. (.not. li_mask_is_ice(cellMask(cell2)))) ) ) then
```
Here, it is only necessary to check which cell is the grounded one:
```suggestion
            if (li_mask_is_grounded_ice(cellMask(cell1))) then
```

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2271,4 +2303,96 @@ subroutine calc_hydro_mask(domain)
+                 ((bedTopography(cell2) < config_sea_level) .and. (.not. li_mask_is_ice(cellMask(cell2)))) ) ) then
+               totalGroundingLineDischargeEdge(iEdge) = abs(channelDischarge(iEdge)) +abs( waterFlux(iEdge) * dvEdge(iEdge))
+               totalGroundingLineDischargeCell(cell2) = totalGroundingLineDischargeCell(cell2) + totalGroundingLineDischargeEdge(iEdge)
+            elseif ( (li_mask_is_grounded_ice(cellMask(cell2))) .and. &
+                     (li_mask_is_floating_ice(cellMask(cell1)) .or. &
+                     ((bedTopography(cell1) < config_sea_level) .and. (.not. li_mask_is_ice(cellMask(cell1)))) ) ) then
```
similar change to previous

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -263,7 +267,7 @@
                 <var name="channelDiffusivity" type="real" dimensions="nEdges Time" units="m^{2} s^{-1}"
                      description="diffusivity in channel in subglacial hydrology system" />
-
+		
```
remove this whitespace change

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -230,7 +230,11 @@
-                <!-- channel variables -->
+	     	<var name="totalGroundingLineDischargeCell" type="real" dimensions="Time nCells" units="m^{3} s^{-1}"
+		     description="total discharge across the grounding line, extrapolated from edge to adjacent ungrounded cell" />	
+	     	<var name="totalGroundingLineDischargeEdge" type="real" dimensions="Time nEdges" units="m^{3} s^{-1}"
+		     description="total discharge across grounding line edge" />
```
Is this field necessary?  I don't remember what we decided with the ocean folks.  I know we want the cell version, but I don't remember if there is potential value to also having the edge version.

matthewhoffman reviewed (changes requested):
@alexolinhager , thanks for creating this PR.  I have a few minor changes to request.

I also had a question of if we actually need the edge version of the GL flux - I'm not sure of the answer.  I was also wondering if while you're doing this if we should create distributed and channel versions in case we ever want to handle them differently in the ocean model in the future.  Or we could wait until we need that.

alexolinhager commented:
@matthewhoffman I added two more commits: The first addresses your comments cleaning up the code and the second calculates the distributed and channelized versions of `totalGroundingLineDischargeCell`, as you suggested. I figure we should add them to the code now instead of wishing we had run with those variables later on. I also removed totalGroundingLineDischargeEdge from the registry because it won't be used by MPAS-Ocean

alexolinhager commented:
The PR also passes the full integration suite so should be good to merge if you give it your stamp of approval

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -229,8 +229,14 @@
-                <!-- channel variables -->
+                        description="time step used for evolving subglacial hydrology system" />
+	     	<var name="distGroundingLineDischargeCell" type="real" dimensions="Time nCells" units="m^{3} s^{-1}"
+		     description="distributed discharge across the grounding line, extrapolated from edge to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
```
```suggestion
		     description="distributed discharge across the grounding line, summed from grounding line edges to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
```

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -229,8 +229,14 @@
+	     	<var name="distGroundingLineDischargeCell" type="real" dimensions="Time nCells" units="m^{3} s^{-1}"
+		     description="distributed discharge across the grounding line, extrapolated from edge to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
+	     	<var name="totalGroundingLineDischargeCell" type="real" dimensions="Time nCells" units="m^{3} s^{-1}"
+		     description="total (channel + dist.) discharge across the grounding line, extrapolated from edge to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
```
```suggestion
		     description="total (channel + dist.) discharge across the grounding line, summed from grounding line edges to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
```

matthewhoffman commented on components/mpas-albany-landice/src/Registry_subglacial_hydro.xml:
```diff
@@ -229,8 +229,14 @@
+		     description="total (channel + dist.) discharge across the grounding line, extrapolated from edge to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
+	     	<!-- channel variables -->
+	     	<var name="chnlGroundingLineDischargeCell" type="real" dimensions="Time nCells" units="m^{3} s^{-1}"
+		     description="channel discharge across the grounding line, extrapolated from edge to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
```
```suggestion
		     description="channel discharge across the grounding line, summed from grounding line edges to adjacent ungrounded cell. Values from all edges are summed if multiple grounding line edges border a single ungrounded cell" />	
```

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -551,6 +551,21 @@ subroutine li_SGH_solve(domain, err)
+
+         block => block % next
+      end do
+      call mpas_timer_start("halo updates")
+      call mpas_dmpar_field_halo_exch(domain, 'totalGroundingLineDischargeCell')
+      call mpas_dmpar_field_halo_exch(domain, 'totalGroundingLineDischargeEdge')
+      call mpas_timer_stop("halo updates")
```
There isn't a halo update on the new variables you added on Friday and edge-version variable is still here, which I would expect would cause an error.  But now that I'm thinking about it, we don't actually need halo updates on these variables.  The values in the halo regions will be incorrect, but the fields are not used for any internal calculations, and the halo regions don't get used when the fields get written to output.  So I think the best thing is just delete these halo updates.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2271,4 +2277,105 @@ subroutine calc_hydro_mask(domain)
+      real (kind=RKIND) :: distGroundingLineDischargeEdge
+      real (kind=RKIND) :: chnlGroundingLineDischargeEdge
+      real (kind=RKIND) :: totalGroundingLineDischargeEdge
+      real (kind=RKIND), pointer :: config_sea_level
```
remove this unused variable here and below at 2350

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2271,4 +2277,105 @@ subroutine calc_hydro_mask(domain)
+      real (kind=RKIND), dimension(:), pointer :: distGroundingLineDischargeCell
+      real (kind=RKIND), dimension(:), pointer :: chnlGroundingLineDischargeCell
+      real (kind=RKIND), dimension(:), pointer :: totalGroundingLineDischargeCell
+      real (kind=RKIND), dimension(:), pointer :: bedTopography
```
remove this unused variable here and below

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -2271,4 +2277,105 @@ subroutine calc_hydro_mask(domain)
+      totalGroundingLineDischargeCell(:) = 0.0_RKIND
+      
+      do iEdge = 1, nEdgesSolve
+         cell1 = cellsOnEdge(1, iEdge)
+         cell2 = cellsOnEdge(2, iEdge)
```
would be more slightly efficient to move these inside the if-statement

matthewhoffman reviewed (changes requested):
@alexolinhager , thanks for revising this.  I have requested a few more small changes.

matthewhoffman commented on components/mpas-albany-landice/src/mode_forward/mpas_li_subglacial_hydro.F:
```diff
@@ -1799,9 +1802,8 @@ subroutine update_channel(block, err)
       call mpas_pool_get_array(hydroPool, 'channelDiffusivity', channelDiffusivity)
       call mpas_pool_get_array(geometryPool, 'edgeMask', edgeMask)
-
+      
```
```suggestion

```

matthewhoffman reviewed (approved):
@alexolinhager , I'm still seeing some whitespace changes, but it's not that important and I don't want to hold up this PR any longer.  Other than that, everything looks good.  I'll get this merged.

matthewhoffman commented:
# Testing
All tests pass `full_integration` on Chicoma
